README.md
# UcloudStorage

Simple API for ucloud storage

## Installation

Add this line to your application's Gemfile:

    gem 'ucloud_storage'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install ucloud_storage

## Usage

    ucloud = UcloudStorage.new(user: 'email', pass: 'API_KEY')

    ucloud.authoize
    # Upload
    ucloud.upload(filepath, boxname, destination)
    # Delete
    ucloud.delete(boxname, destination)
    # Get
    ucloud.get(boxname, destination)

## Configuration

Set default user/pass info

    UcloudStorage.configure do |config|
      config.user = 'email'
      config.pass = 'API KEY'
      # config.type = 'standard'
    end

## Response block

Every request yields a response

    ucloud.upload(filepath, boxname, dest) do |response|
      response_code = response.code
    end


## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


CHANGELOG
## v0.0.15

* support ucloud lite authentication
* fix uploaded to double slashed destination url (http://ucloud.com/box_name//destination)
* refactor specs
* remove uploaded file after test

## v0.0.14

* fix bug - block not called

## v0.0.13

* retry when authorizion expire

## v0.0.12

* remove fork

## v0.0.2

* Add runtime_dependency

## v0.0.2

* Add Module

## v0.0.1

* Initial Release

spec/support/vcr.rb
VCR.configure do |c|
	c.cassette_library_dir = File.join(File.dirname(__FILE__), "../fixtures/vcr_cassettes")
  c.allow_http_connections_when_no_cassette = true
	c.hook_into :webmock
end

spec/ucloud_storage_spec.rb
# encoding: utf-8
require_relative '../lib/ucloud_storage'
require 'httparty'
require 'vcr'
require 'support/vcr'
require 'yaml'

describe UcloudStorage do
  context "set user/pass info" do
    before do
      UcloudStorage.configure do |config|
        config.user = "abc@mintshop.com"
        config.pass = "my_api_key"
      end
    end

    it "can set default user/pass info" do
      UcloudStorage.user.should == "abc@mintshop.com"
      UcloudStorage.pass.should == "my_api_key"
    end

    it 'uses default uesr/pass info when initialized' do
      ucloud = UcloudStorage.new
      ucloud.user.should == "abc@mintshop.com"
      ucloud.pass.should == "my_api_key"
    end
  end

  let(:valid_ucloud) do
    file = File.open(File.join(File.dirname(__FILE__), "/support/auth_info.yml"))
    auth_info = YAML.load(file)
    ucloud = UcloudStorage.new(user: auth_info["valid_user"]["user"],
                               pass: auth_info["valid_user"]["pass"])
  end

  let(:valid_ucloud_lite) do
    file = File.open(File.join(File.dirname(__FILE__), "/support/auth_info.yml"))
    auth_info = YAML.load(file)
    ucloud = UcloudStorage.new(user: auth_info["valid_user"]["user"],
                               pass: auth_info["valid_user"]["pass"],
                               type: 'lite')
  end

  let(:invalid_ucloud) do
    invlaid_ucloud = UcloudStorage.new
    invlaid_ucloud.user = "invalid_user@mintshop.com"
    invlaid_ucloud.pass = "please download mintshop"
    invlaid_ucloud
  end

  let(:file_path) { File.join(File.dirname(__FILE__), "/fixtures/sample_file.txt") }
  let(:box) { 'dev_box' }
  let(:destination) { 'cropped_images/'+Pathname(file_path).basename.to_s }

  describe '#authorize' do
    it "can authorize with valid user/pass" do
      VCR.use_cassette("storage/v1/auth") do
        valid_ucloud.is_authorized?.should_not == true
        valid_ucloud.authorize.should == true
        valid_ucloud.storage_url.should_not be_nil
        valid_ucloud.auth_token.should_not be_nil
        valid_ucloud.is_authorized?.should == true
      end
    end

    it 'should be authorize with lite storage' do
      VCR.use_cassette("storage/v1/authlite") do
        valid_ucloud_lite.is_authorized?.should_not == true
        valid_ucloud_lite.authorize.should == true
        valid_ucloud_lite.storage_url.should_not be_nil
        valid_ucloud_lite.auth_token.should_not be_nil
        valid_ucloud_lite.is_authorized?.should == true
      end
    end

    it "cannot authorize with invalid user/pass" do
      VCR.use_cassette("storage/v1/auth_fail") do
        invalid_ucloud.authorize.should == false
        invalid_ucloud.storage_url.should be_nil
        invalid_ucloud.auth_token.should be_nil
      end
    end

    it 'yields response' do
      VCR.use_cassette("storage/v1/auth") do
        valid_ucloud.authorize do |response|
          response.code.should == 200
        end
      end
    end
  end

  describe "#upload" do
    it "can upload a file" do
      VCR.use_cassette("v1/put_storage_object") do
        valid_ucloud.authorize
        valid_ucloud.upload(file_path, box, destination).should be_true
        uploaded_url = "https://ssproxy.ucloudbiz.olleh.com/v1/AUTH_f46e842e-c688-460e-a70b-e6a4d30e9885/dev_box/cropped_images/sample_file.txt"
        HTTParty.get(uploaded_url).header.code_type.should == Net::HTTPOK
        valid_ucloud.delete(box, destination)
      end
    end

    it "can upload a file with proper url" do
      VCR.use_cassette("v1/put_storage_object_02") do
        valid_ucloud.authorize
        destination_with_slash = "/" + destination
        valid_ucloud.upload(file_path, box, destination_with_slash).should be_true
        uploaded_url = "https://ssproxy.ucloudbiz.olleh.com/v1/AUTH_f46e842e-c688-460e-a70b-e6a4d30e9885/dev_box/cropped_images/sample_file.txt"
        HTTParty.get(uploaded_url).header.code_type.should == Net::HTTPOK
        valid_ucloud.delete(box, destination)
      end
    end

    it "should fail to upload with invalid file path" do
      VCR.use_cassette('storage/v1/auth') do
        valid_ucloud.authorize
        invalid_file_path = File.join(File.dirname(__FILE__), "/fixtures/no_sample_file.txt")
        expect {
          valid_ucloud.upload(invalid_file_path, box, destination)
        }.to raise_error(Errno::ENOENT)
      end
    end

    it "should fail to upload without authorization" do
      expect {
        valid_ucloud.upload(file_path, box, destination).should be_true
      }.to raise_error(UcloudStorage::NotAuthorized)
    end

    it "should retry to upload if authorization failure response" do
      VCR.use_cassette("v1/put_storage_object_with_auth_fail") do
        valid_ucloud.authorize
        valid_ucloud.auth_token += "a"

        valid_ucloud.upload(file_path, box, destination) do |response|
          response.code.should == 201
        end
        valid_ucloud.delete(box, destination)
      end
    end

    it 'yields response' do
      VCR.use_cassette("v1/put_storage_object") do
        valid_ucloud.authorize
        valid_ucloud.upload(file_path, box, destination) do |response|
          response.code.should == 201
        end
        valid_ucloud.delete(box, destination)
      end
    end
  end

  describe "#delete" do
    it 'should delete updated object' do
      VCR.use_cassette("v1/delete_storage_object_02") do
        valid_ucloud.authorize
        valid_ucloud.upload(file_path, box, destination)

        valid_ucloud.delete(box, destination) do |response|
          response.code.should == 204
        end.should == true
        uploaded_url = "https://ssproxy.ucloudbiz.olleh.com/v1/AUTH_f46e842e-c688-460e-a70b-e6a4d30e9885/dev_box/cropped_images/sample_file.txt"
        HTTParty.get(uploaded_url).header.code_type.should == Net::HTTPNotFound
      end
    end
  end

  describe "#exist" do
    it "should retry to upload if authorization failure response" do
      VCR.use_cassette("v1/object_exists") do
        valid_ucloud.authorize
        valid_ucloud.upload(file_path, box, destination) do |response|
          response.code.should == 201
        end

        valid_ucloud.auth_token += "a"

        valid_ucloud.get(box, destination) do |response|
          [200, 304].should include(response.code)
        end.should == true
      end
    end

    it 'should get updated object' do
      VCR.use_cassette("v1/get_updated_object") do
        valid_ucloud.authorize
        valid_ucloud.upload(file_path, box, destination)

        result = false
        valid_ucloud.get(box, destination) do |response|
          result = true
          [200, 304].should include(response.code)
        end
        result.should == true
        valid_ucloud.delete(box, destination)
      end
    end
  end
end


lib/ucloud_storage/ucloud_storage.rb
# encoding: utf-8
require 'httparty'

module UcloudStorage
  class TotalyWrongException < StandardError; end
  class NotAuthorized < StandardError; end

  class UcloudStorage
    attr_accessor :user, :pass, :type, :storage_url, :auth_token

    def initialize(options={})
      @user = options.fetch(:user) { Configuration.user }
      @pass = options.fetch(:pass) { Configuration.pass }
      @type = options.fetch(:type) { Configuration.type || 'standard' }
      @authorized = false
    end

    def authorize
      response = HTTParty.get(auth_url, headers: { "X-Storage-User" => user,
                                                   "X-Storage-Pass" => pass })

      yield response if block_given?

      case response.code
      when 200
        self.storage_url = response.headers["X-Storage-Url"]
        self.auth_token = response.headers["X-Auth-Token"]
        @authorized = true
      when 401 then false
      else raise TotalyWrongException
      end
    end

    def is_authorized?
      @authorized
    end

    def upload(file_path, box_name, destination, &block)
      raise NotAuthorized if storage_url.nil?

      file = File.new(file_path)
      content_type = get_content_type(file_path)
      upload_blob(file.read, box_name, destination, content_type, &block)
    end

    def upload_blob(blob, box_name, destination, content_type, &block)
      raise NotAuthorized if storage_url.nil?

      target_url = File.join(storage_url, box_name, destination).to_s
      response = HTTParty.put(target_url,
                              headers: {
                                "X-Auth-Token" => auth_token,
                                "Content-Type" => content_type,
                                "Content-Length" => blob.size.to_s },
                              body: blob)

      if response.code == 401 and authorize
        return upload_blob(blob, box_name, destination, content_type, &block)
      end

      yield response if block_given?

      response.code == 201 ? true : false
    end

    def delete(box_name, destination, &block)
      request(:delete, box_name, destination, [204], &block)
    end

    def get(box_name, destination, &block)
      request(:get, box_name, destination, [200, 304], &block)
    end

    def exist?(box_name, destination, &block)
      request(:head, box_name, destination, [200, 204], &block)
    end

    private

    def auth_url
      case type
      when "standard"
        "https://api.ucloudbiz.olleh.com/storage/v1/auth"
      when 'standard-jpn'
        "https://api.ucloudbiz.olleh.com/storage/v1/authjp"
      when "lite"
        "https://api.ucloudbiz.olleh.com/storage/v1/authlite"
      end
    end

    def request(method, box_name, destination, success_code = [200], &block)
      raise NotAuthorized if storage_url.nil?

      target_url = File.join(storage_url, box_name, destination)
      response = HTTParty.send(method, target_url, headers: { "X-Auth-Token" => auth_token })
      return request(method, box_name, destination, success_code) if response.code == 401 and authorize

      yield response if block_given?
      success_code.include?(response.code) ? true : false
    end

    #  stolen from http://stackoverflow.com/a/16636012/1802026
    def get_content_type(local_file_path)
      png = Regexp.new("\x89PNG".force_encoding("binary"))
      jpg = Regexp.new("\xff\xd8\xff\xe0\x00\x10JFIF".force_encoding("binary"))
      jpg2 = Regexp.new("\xff\xd8\xff\xe1(.*){2}Exif".force_encoding("binary"))
      case IO.read(local_file_path, 10)
      when /^GIF8/
        'image/gif'
      when /^#{png}/
        'image/png'
      when /^#{jpg}/
        'image/jpeg'
      when /^#{jpg2}/
        'image/jpeg'
      else
        if local_file_path.end_with? '.txt'
          'text/plain'
        else
          'application/octet-stream'
        end
        # mime_type = `file #{local_file_path} --mime-type`.gsub("\n", '') # Works on linux and mac
        # raise UnprocessableEntity, "unknown file type" if !mime_type
        # mime_type.split(':')[1].split('/')[1].gsub('x-', '').gsub(/jpeg/, 'jpg').gsub(/text/, 'txt').gsub(/x-/, '')
      end
    end

  end
end
